README.md

```diff
@@ -65,178 +68,106 @@ Giskard works with any model, in any environment and integrates seamlessly with
+climate_qa_chain = RetrievalQA.from_llm(llm=llm, retriever=db.as_retriever(), prompt=prompt)
+```
 
-# Replace this with your own data & model creation.
-df = giskard.demo.titanic_df()
-demo_data_processing_function, demo_sklearn_model = giskard.demo.titanic_pipeline()
+## 2. 🔎 Scan your model for issues
 
-# Wrap your Pandas DataFrame with Giskard.Dataset (test set, a golden dataset, etc.).
-giskard_dataset = giskard.Dataset(
-    df=df,  # A pandas.DataFrame that contains the raw data (before all the pre-processing steps) and the actual ground truth variable (target).
-    target="Survived",  # Ground truth variable
-    name="Titanic dataset", # Optional
-    cat_columns=['Pclass', 'Sex', "SibSp", "Parch", "Embarked"]  # List of categorical columns. Optional, but is a MUST if available. Inferred automatically if not.
-)
+Next, wrap your agent to prepare it for Giskard's scan:
+
+```python
+import giskard
+import pandas as pd
+
+def model_predict(df: pd.DataFrame):
+    """Wraps the LLM call in a simple Python function.
 
-# Wrap your model with Giskard.Model. Check the dedicated doc page: https://docs.giskard.ai/en/latest/guides/wrap_model/index.html
-# you can use any tabular, text or LLM models (PyTorch, HuggingFace, LangChain, etc.),
-# for classification, regression & text generation.
-def prediction_function(df):
-    # The pre-processor can be a pipeline of one-hot encoding, imputer, scaler, etc.
-    preprocessed_df = demo_data_processing_function(df)
-    return demo_sklearn_model.predict_proba(preprocessed_df)
+    The function takes a pandas.DataFrame containing the input variables needed
+    by your model, and must return a list of the outputs (one for each row).
+    """
+    return [climate_qa_chain.run({"query": question}) for question in df["question"]]
 
+# Don’t forget to fill the `name` and `description`: they are used by Giskard
+# to generate domain-specific tests.
 giskard_model = giskard.Model(
-    model=prediction_function,  # A prediction function that encapsulates all the data pre-processing steps and that could be executed with the dataset used by the scan.
-    model_type="classification",  # Either regression, classification or text_generation.
-    name="Titanic model",  # Optional
-    classification_labels=demo_sklearn_model.classes_,  # Their order MUST be identical to the prediction_function's output order
-    feature_names=['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Embarked'],  # Default: all columns of your dataset
+    model=model_predict,
+    model_type="text_generation",
+    name="Climate Change Question Answering",
+    description="This model answers any question about climate change based on IPCC reports",
+    feature_names=["question"],
 )
 ```
 
 ✨✨✨Then run Giskard's magical scan✨✨✨
 ```python
-scan_results = giskard.scan(giskard_model, giskard_dataset)
+scan_results = giskard.scan(giskard_model)
```

Comment:
@luca-martial can't run this cell now

'''
{
	"name": "BadRequestError",
	"message": "Error code: 400 - {'error': {'message': \"Invalid parameter: 'response_format' of type 'json_object' is not supported with this model.\", 'type': 'invalid_request_error', 'param': 'response_format', 'code': None}}",
	"stack": "---------------------------------------------------------------------------
BadRequestError                           Traceback (most recent call last)
Cell In[19], line 1
----> 1 scan_results = giskard.scan(giskard_model)
      2 scan_results.to_html(\"Luca_scan_results.html\")

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/giskard/scanner/__init__.py:64, in scan(model, dataset, features, params, only, verbose, raise_exceptions)
     35 \"\"\"Automatically detects model vulnerabilities.
     36 
     37 See :class:`Scanner` for more details.
   (...)
     61     A scan report object containing the results of the scan.
     62 \"\"\"
     63 scanner = Scanner(params, only=only)
---> 64 return scanner.analyze(
     65     model, dataset=dataset, features=features, verbose=verbose, raise_exceptions=raise_exceptions
     66 )

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/giskard/scanner/scanner.py:100, in Scanner.analyze(self, model, dataset, features, verbose, raise_exceptions)
     77 \"\"\"Runs the analysis of a model and dataset, detecting issues.
     78 
     79 Parameters
   (...)
     96     A report object containing the detected issues and other information.
     97 \"\"\"
     98 with TemporaryRootLogLevel(logging.INFO if verbose else logging.NOTSET):
     99     # Check that the model and dataset were appropriately wrapped with Giskard
--> 100     model, dataset, model_validation_time = self._validate_model_and_dataset(model, dataset)
    102     # Check that provided features are valid
    103     features = self._validate_features(features, model, dataset)

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/giskard/scanner/scanner.py:209, in Scanner._validate_model_and_dataset(self, model, dataset)
    203 if dataset is not None and not isinstance(dataset, Dataset):
    204     raise ValueError(
    205         \"The dataset object you provided is not valid. Please wrap your dataframe with `giskard.Dataset`. \"
    206         \"You can follow the docs here: https://docs.giskard.ai/en/latest/guides/wrap_dataset/index.html\"
    207     )
--> 209 model, dataset = self._prepare_model_dataset(model, dataset)
    211 if not model.is_text_generation:
    212     time_start = perf_counter()

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/giskard/scanner/scanner.py:268, in Scanner._prepare_model_dataset(self, model, dataset)
    266 logger.debug(\"Automatically generating test dataset.\")
    267 try:
--> 268     return model, generate_test_dataset(model)
    269 except LLMGenerationError:
    270     warning(
    271         \"Failed to generate test dataset. Trying to run the scan with an empty dataset. For improved results, please provide a test dataset.\"
    272     )

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/giskard/llm/utils.py:43, in generate_test_dataset(model, num_samples, temperature, column_types, llm_seed)
     12 \"\"\"Generates a synthetic test dataset using an LLM.
     13 
     14 Parameters
   (...)
     39 :class:`giskard.llm.generators.BaseDataGenerator`
     40 \"\"\"
     41 generator = SimpleDataGenerator(llm_temperature=temperature, llm_seed=llm_seed)
---> 43 return generator.generate_dataset(model, num_samples, column_types)

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/giskard/llm/generators/base.py:64, in _BaseLLMGenerator.generate_dataset(self, model, num_samples, column_types)
     41 \"\"\"Generates a test dataset for the model.
     42 
     43 Parameters
   (...)
     60     If the generation fails.
     61 \"\"\"
     62 messages = self._format_messages(model, num_samples, column_types)
---> 64 out = self.llm_client.complete(
     65     messages=messages,
     66     temperature=self.llm_temperature,
     67     caller_id=self.__class__.__name__,
     68     seed=self.llm_seed,
     69     format=\"json\",
     70 )
     72 generated = self._parse_output(out)
     74 dataset = Dataset(
     75     df=pd.DataFrame(generated),
     76     name=self._make_dataset_name(model),
     77     validation=False,
     78     column_types=column_types,
     79 )

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/giskard/llm/client/openai.py:49, in OpenAIClient.complete(self, messages, temperature, max_tokens, caller_id, seed, format)
     46     extra_params[\"response_format\"] = {\"type\": \"json_object\"}
     48 try:
---> 49     completion = self._client.chat.completions.create(
     50         model=self.model,
     51         messages=[asdict(m) for m in messages],
     52         temperature=temperature,
     53         max_tokens=max_tokens,
     54         **extra_params,
     55     )
     56 except openai.AuthenticationError as err:
     57     raise LLMConfigurationError(AUTH_ERROR_MESSAGE) from err

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/sentry_sdk/integrations/openai.py:181, in _wrap_chat_completion_create.<locals>.new_chat_completion(*args, **kwargs)
    179     _capture_exception(Hub.current, e)
    180     span.__exit__(None, None, None)
--> 181     raise e from None
    183 with capture_internal_exceptions():
    184     if _should_send_default_pii() and integration.include_prompts:

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/sentry_sdk/integrations/openai.py:177, in _wrap_chat_completion_create.<locals>.new_chat_completion(*args, **kwargs)
    175 span.__enter__()
    176 try:
--> 177     res = f(*args, **kwargs)
    178 except Exception as e:
    179     _capture_exception(Hub.current, e)

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/openai/_utils/_utils.py:275, in required_args.<locals>.inner.<locals>.wrapper(*args, **kwargs)
    273             msg = f\"Missing required argument: {quote(missing[0])}\"
    274     raise TypeError(msg)
--> 275 return func(*args, **kwargs)

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/openai/resources/chat/completions.py:667, in Completions.create(self, messages, model, frequency_penalty, function_call, functions, logit_bias, logprobs, max_tokens, n, presence_penalty, response_format, seed, stop, stream, temperature, tool_choice, tools, top_logprobs, top_p, user, extra_headers, extra_query, extra_body, timeout)
    615 @required_args([\"messages\", \"model\"], [\"messages\", \"model\", \"stream\"])
    616 def create(
    617     self,
   (...)
    665     timeout: float | httpx.Timeout | None | NotGiven = NOT_GIVEN,
    666 ) -> ChatCompletion | Stream[ChatCompletionChunk]:
--> 667     return self._post(
    668         \"/chat/completions\",
    669         body=maybe_transform(
    670             {
    671                 \"messages\": messages,
    672                 \"model\": model,
    673                 \"frequency_penalty\": frequency_penalty,
    674                 \"function_call\": function_call,
    675                 \"functions\": functions,
    676                 \"logit_bias\": logit_bias,
    677                 \"logprobs\": logprobs,
    678                 \"max_tokens\": max_tokens,
    679                 \"n\": n,
    680                 \"presence_penalty\": presence_penalty,
    681                 \"response_format\": response_format,
    682                 \"seed\": seed,
    683                 \"stop\": stop,
    684                 \"stream\": stream,
    685                 \"temperature\": temperature,
    686                 \"tool_choice\": tool_choice,
    687                 \"tools\": tools,
    688                 \"top_logprobs\": top_logprobs,
    689                 \"top_p\": top_p,
    690                 \"user\": user,
    691             },
    692             completion_create_params.CompletionCreateParams,
    693         ),
    694         options=make_request_options(
    695             extra_headers=extra_headers, extra_query=extra_query, extra_body=extra_body, timeout=timeout
    696         ),
    697         cast_to=ChatCompletion,
    698         stream=stream or False,
    699         stream_cls=Stream[ChatCompletionChunk],
    700     )

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/openai/_base_client.py:1213, in SyncAPIClient.post(self, path, cast_to, body, options, files, stream, stream_cls)
   1199 def post(
   1200     self,
   1201     path: str,
   (...)
   1208     stream_cls: type[_StreamT] | None = None,
   1209 ) -> ResponseT | _StreamT:
   1210     opts = FinalRequestOptions.construct(
   1211         method=\"post\", url=path, json_data=body, files=to_httpx_files(files), **options
   1212     )
-> 1213     return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/openai/_base_client.py:902, in SyncAPIClient.request(self, cast_to, options, remaining_retries, stream, stream_cls)
    893 def request(
    894     self,
    895     cast_to: Type[ResponseT],
   (...)
    900     stream_cls: type[_StreamT] | None = None,
    901 ) -> ResponseT | _StreamT:
--> 902     return self._request(
    903         cast_to=cast_to,
    904         options=options,
    905         stream=stream,
    906         stream_cls=stream_cls,
    907         remaining_retries=remaining_retries,
    908     )

File ~/Code/internal-red-teaming/.venv/lib/python3.11/site-packages/openai/_base_client.py:993, in SyncAPIClient._request(self, cast_to, options, remaining_retries, stream, stream_cls)
    990         err.response.read()
    992     log.debug(\"Re-raising status error\")
--> 993     raise self._make_status_error_from_response(err.response) from None
    995 return self._process_response(
    996     cast_to=cast_to,
    997     options=options,
   (...)
   1000     stream_cls=stream_cls,
   1001 )

BadRequestError: Error code: 400 - {'error': {'message': \"Invalid parameter: 'response_format' of type 'json_object' is not supported with this model.\", 'type': 'invalid_request_error', 'param': 'response_format', 'code': None}}"
}
'''